backend/src/socket/socket.gateway.ts

```diff
@@ -26,187 +26,265 @@ import { Game, Player } from '../dto/game.dto';
+    const existingPlayer: string = await this.cacheManager.get(
+      `game${jwtData.email}`,
+    );
+    if (existingPlayer) {
+      let existingPlayerObject: Player = JSON.parse(existingPlayer);
+      existingPlayerObject.socketID = client.id;
+      await this.cacheManager.set(
+        `game${jwtData.email}`,
+        JSON.stringify(existingPlayerObject),
+      );
+      console.log(existingPlayerObject);
+      console.log(
+        'Socket: existing player updated: ',
+        existingPlayerObject.email,
+      );
+      return;
+    }
 
-			const newPlayer = new Player(null, jwtData.email, client.id, user.userName, 0);
-			this.clients.push(newPlayer);
-			console.log('creating new client/player:', newPlayer.email);
-		}
-	}
+    // find the user in database
+    const user = await this.prisma.user.findUnique({
+      where: {
+        email: jwtData.email,
+      },
+    });
 
-  async handleDisconnect(client: Socket) {
-    const existingClient = this.clients.find((c) => c.socketID === client.id);
-    if (existingClient){
-      const user = await this.prisma.user.findUnique({
-        where: {
-          email: existingClient.email,
-        },
+    if (!user) {
+      client.emit('accountDeleted', {
+        message: 'Your account has been deleted.',
       });
-      if (user.status === 'OFFLINE'){
-        this.clients = this.clients.filter((c) => c.email !== existingClient.email);
-				return ;
-      } else if (user.status === 'WAITING'){
-        const pendingPlayer: string = await this.cacheManager.get('pendingPlayer');
-        if (pendingPlayer){
-          const pendingPlayerObject: Player = JSON.parse(pendingPlayer);
-          if (user.email === pendingPlayerObject.email){
-            await this.cacheManager.del('pendingPlayer');
-       		}
-      	}} else if (user.status === 'PLAYING') {
-					// wait 30 seconds, or game is lost??
-					if (existingClient.gameID){
-						client.join(existingClient.gameID.toString())
-					}
-				}
-    	}
-    console.log(client.id, ' disconnected from generic socket. 0.0');
+      client.disconnect();
+      return;
+    }
+
+    // if no such player in redis, create one.
+    const newPlayer = new Player(
+      'non',
+      jwtData.email,
+      client.id,
+      user.userName,
+      0,
+    );
+    // save the new Player in redis
+    await this.cacheManager.set(
+      `game${jwtData.email}`,
+      JSON.stringify(newPlayer),
+    );
+    console.log('Socket: new client/player created:', newPlayer.email);
+  }
+
+  async handleDisconnect(client: Socket) {
+    // check if player left in game
+    const playerEmail: string = await this.cacheManager.get(client.id);
+    if (!playerEmail) return;
+    await this.cacheManager.del(client.id);
+
+    const existingPlayer: string = await this.cacheManager.get(
+      `game${playerEmail}`,
+    );
+    if (!existingPlayer) return;
+    const existingPlayerObject: Player = JSON.parse(existingPlayer);
```

Comment:
Why do we add separate entries in the cache for socket and player? Wouldn't it be better to keep all info in one place?